ProcessList.c

```diff
@@ -165,22 +179,18 @@ void ProcessList_add(ProcessList* this, Process* p) {
    Vector_add(this->processes, p);
    Hashtable_put(this->processTable, p->pid, p);
 
-   assert(Vector_indexOf(this->processes, p, Process_pidCompare) != -1);
+   assert(ProcessList_indexOf(this, p) != -1);
    assert(Hashtable_get(this->processTable, p->pid) != NULL);
    assert(Hashtable_count(this->processTable) == Vector_count(this->processes));
 }
 
-void ProcessList_remove(ProcessList* this, const Process* p) {
-   assert(Vector_indexOf(this->processes, p, Process_pidCompare) != -1);
-   assert(Hashtable_get(this->processTable, p->pid) != NULL);
-
-   const Process* pp = Hashtable_remove(this->processTable, p->pid);
-   assert(pp == p); (void)pp;
-
+static void ProcessList_removeIndex(ProcessList* this, const Process* p, int idx) {
    pid_t pid = p->pid;
-   int idx = Vector_indexOf(this->processes, p, Process_pidCompare);
-   assert(idx != -1);
 
+   assert(((Process*)Vector_get(this->processes, idx))->pid == pid);
+   assert(Hashtable_get(this->processTable, pid) != NULL);
+
+   Hashtable_remove(this->processTable, pid);
    if (idx >= 0) {
       Vector_remove(this->processes, idx);
    }
```

Comment:
```suggestion
   if (idx >= 0 && idx < Vector_size(this->processes))
      Vector_remove(this->processes, idx);

```

Alternative, make the index check an assertion and unconditionally remove the Vector entry.

Reply:
Good catch, there's an assertion before this that will fail if `idx` is out of range ([`devname.c#L103-L114`](https://github.com/apple-oss-distributions/Libc/blob/Libc-1439.141.1/gen/devname.c#L103-L114)) so we're good here. But the public function `ProcessList_remove` was not checking if the `Process` was found before calling `ProcessList_removeIndex` so I changed it to check before attempting to remove it. IMHO if `Vector_indexOf` returns a value that is past the end of the vector then we have bigger problems and a segfault might be appropriate.

```c
void ProcessList_remove(ProcessList* this, const Process* p) {
   int idx = Vector_indexOf(this->processes, p, Process_pidEqual);
   assert(idx >= 0);
   if (idx >= 0) {
      ProcessList_removeIndex(this, p, idx);
      // This is public function so do not require
      // callers to compact the Vector
      Vector_compact(this->processes);
   }
}
```

Reply:
Unfortunately an index just off the end of the Vector won't guarantee a SIGSEGV …